src/unix/udp.c

```diff
@@ -654,28 +654,71 @@ int uv__udp_connect(uv_udp_t* handle,
   return 0;
 }
 
-
+/* From https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html
+ * Any of uv supported UNIXs kernel should be standardized, but the kernel 
+ * implementation logic not same, let's use pseudocode to explain the udp
+ * disconnect behaviors:
+ * 
+ * Predefined stubs for pseudocode:
+ *   1. sodisconnect: The function to perform the real udp disconnect
+ *   2. pru_connect: The function to perform the real udp connect
+ *   3. so: The kernel object match with socket fd
+ *   4. addr: The sockaddr parameter from user space
+ * 
+ * BSDs:
+ *   if(sodisconnect(so) == 0) { // udp disconnect succeed
+ *     if (addr->sa_len != so->addr->sa_len) return EINVAL;
+ *     if (addr->sa_family != so->addr->sa_family) return EAFNOSUPPORT;
+ *     pru_connect(so);
+ *   }
+ *   else return EISCONN;
+ *
+ * z/OS (same with Windows):
+ *   if(addr->sa_len < so->addr->sa_len) return EINVAL;
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ *
+ * AIX:
+ *   if(addr->sa_len != sizeof(struct sockaddr)) return EINVAL; // ignore ip proto version
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ *
+ * Linux,Others:
+ *   if(addr->sa_len < sizeof(struct sockaddr)) return EINVAL;
+ *   if (addr->sa_family == AF_UNSPEC) sodisconnect(so);
+ */
 int uv__udp_disconnect(uv_udp_t* handle) {
     int r;
+#if defined(__MVS__)
+    struct sockaddr_storage addr;
+#else
     struct sockaddr addr;
+#endif
 
     memset(&addr, 0, sizeof(addr));
-
+    
+#if defined(__MVS__)
+    addr.ss_family = AF_UNSPEC;
+#else
     addr.sa_family = AF_UNSPEC;
-
+#endif
+    
     do {
       errno = 0;
-      r = connect(handle->io_watcher.fd, &addr, sizeof(addr));
+      r = connect(handle->io_watcher.fd, (struct sockaddr*)&addr, sizeof(addr));
     } while (r == -1 && errno == EINTR);
 
-    if (r == -1 && errno != EAFNOSUPPORT)
+    if (r == -1) {
+#if defined(BSD) // The macro BSD is from sys/param.h
```

Comment:
We've never used this one before, as normally we write them each out. Are we sure all BSDs have the same behavior here, even though the implementations of Mach, FreeBSD, OpenBSD, etc. tend to have diverged quite a long time ago.

Reply:
Here is the collection of known BSDs have the common behavior:
- [386BSD](https://github.com/386bsd/386bsd/blob/2.0/usr/src/kernel/kern/socket/sockops.c#L261)
- [4.4BSD](https://github.com/chenshuo/4.4BSD-Lite2/blob/master/sys/kern/uipc_socket.c#L252)
- [FreeBSD](https://github.com/freebsd/freebsd-src/blob/main/sys/kern/uipc_socket.c#L1388)
- [DragonFlyBSD](https://github.com/DragonFlyBSD/DragonFlyBSD/blob/master/sys/kern/uipc_socket.c#L698)
- [NetBSD](https://github.com/NetBSD/src/blob/trunk/sys/kern/uipc_socket.c#L830)
- [OpenBSD](https://github.com/openbsd/src/blob/master/sys/kern/uipc_socket.c#L427)
- [PureDarwin](https://github.com/PureDarwin/PureDarwin/blob/main/src/Kernel/xnu/bsd/kern/uipc_socket.c#L1676)
- [MidnightBSD](https://github.com/MidnightBSD/src/blob/master/sys/kern/uipc_socket.c#L1002)
- [GhostBSD](https://github.com/ghostbsd/ghostbsd-src/blob/stable/13/sys/kern/uipc_socket.c#L1325)
- darwin/macOS/iOS
You can see, the logic comes from the first open source BSD operating system `386BSD`,  I think the check enough for libuv to support mostly known BSDs.